### PR TITLE
Add error reporting when debugging on Windows

### LINKED ISSUE #47900: Intellisense on Windows doesn't report real error message by default
## 🐛 Bug

Hi Guys:
I use CMake and the CMakeLists.txt is from pytorch.org , Libtorch 1.7 and Visual studio 2019,
I can output a tensor, it works just fine.
But there is a exception happened, when I simply run backward().
image

CMakeLists.txt I used.(just from pytorch.org )
image

## To Reproduce

1. Download libtorch 1.7 debug cpu. (libtorch-win-shared-with-deps-debug-1.7.0+cpu.zip)
2. Use CMake just like pytorch.org recommended.
3. Run simple backward() function, and get exception.

## Environment

 - Win7 x64
 - Microsoft Visual Studio Professional 2019 version 16.8.0
 - libtorch-win-shared-with-deps-debug-1.7.0+cpu
 - CMake 3.18.4
 

### Anything wants me to pose to make this issue be reproduced and solve is welcome.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @peterjc123 @mszhanyi @skyline75489 @nbcsm @maxluk @guyang3532 @gunandrose4u @smartcat2010

### PR DESCRIPTION
**Pull request by peterjc123**
Fixes #47900

### CODE REVIEW

**Comment by ezyang** [comment]
Nyeh, I don't think we're allowed to do this by default, because the exception might be caught (we *shouldn't* do that but I know at least of one try-catch block). Not sure if making it conditional on there being a debugger attached is enough. Willing to give it a try though, if you think it's worth it.

**Comment by peterjc123** [comment]
@ezyang Okay, will try to find some better alternatives but it is hard since we need to deal with the exception earlier than the debugger.

**Comment by ezyang** [comment]
Yeah I know, the only other possibility is an atexit handler or a top level try-catch (which will be hard to install)

**Comment by peterjc123** [comment]
@ezyang The only chance is to use the Vectored Exception Handlers which is a part of the SEH on Windows. Or, may be just write the exception message into the debug window for debug build which is a simple solution.

**Comment by ezyang** [comment]
Yeah I think I'm ok with spewing exceptions to stderr in debug mode, cuz (1) it's debug mode and (2) we really shouldn't be throwing exceptions except in truly exceptional cases

**Review by malfet — APPROVED** [review]

File: c10/util/Exception.cpp [hunk 1]
```diff
@@ -13,6 +13,25 @@ namespace c10 {
...
+  // the message before it is actually thrown.
+  if (IsDebuggerPresent()) {
+    std::cerr << what_ << std::endl;
+
+    // Tries to bring up the console if it is minimized
+    HWND cw_hwnd = GetConsoleWindow();
+    if (cw_hwnd) {
+      ShowWindow(cw_hwnd, SW_NORMAL);
+    }
+
```
Comment (malfet) [inline c10/util/Exception.cpp:29]:
```suggestion

```

**Pull request closed without merging**